### PR TITLE
Add unit tests for simulation post-process scripts (closes #121)

### LINKED ISSUE #121: Unit tests for klayout_package/python/scripts/simulations/post_process
This issue is part of **unitaryHACK26**. You have to be [registered](https://unitaryhack.dev/register/) to complete this issue.
Learn more about the PR submission process [here](https://unitaryhack.dev/hacker-guide/) and about unitaryHACK rules [here](https://unitary.design/rules/)!

**A note about AI Slop:** while we are open to collaboration with LLMs for unitaryHACK, fully AI-generated PRs are not acceptable. It is at the maintainers' discretion whether or not LLM-generated PRs are the right fit for the issues, and those that appear fully AI-generated may be immediately rejected. Read unitaryHACK's full AI Policy [here](https://unitaryhack.dev/ai-guide/)

---

### Description

Under [klayout_package/python/scripts/simulations/post_process](https://github.com/iqm-finland/KQCircuits/tree/main/klayout_package/python/scripts/simulations/post_process) folder there are several python scripts that are run after simulation execution to perform after-simulation steps, such as compiling simulation results or cleaning up files. These scripts live outside of the KQCircuits core library and are currently not tested. The task is to implement a unit test infrastructure that performs following steps:
- Set up a temporary folder resembling a simulation folder with some "mocked" simulation results
- Run a post process script on this folder, resulting in new files
- Parse files created by post process script to check for validity

The main focus of this task is to implement an infrastrucutre for this type of tests to exists, not so much that these tests are comprehensive and cover all edge cases.

### Prerequisites

> [!IMPORTANT]
> It is best to work on this issue with a working simulator in your device. [Ansys](https://www.ansys.com/) is proprietary, but [Elmer](https://www.elmerfem.org/blog/binaries/) is open source. There are precompiled binaries of Elmer for Windows, but you can also consider installing a Linux image in Windows Sybsystem in Linux ([WSL](https://learn.microsoft.com/en-us/windows/wsl/install)), then installing Elmer from [singularity image](https://iqm-finland.github.io/KQCircuits/developer/containers.html#singularity-usage) we maintain. Unfortunately there is limited Elmer support for mac developers.

Follow the standalone installation guide is [here](https://iqm-finland.github.io/KQCircuits/developer/standalone.html). We recommend ``Reproducible, Secure Installation`` option, command being:

```
cd klayout_package/python
python -m pip install -r requirements/<platform>/requirements.txt
python -m pip install -r requirements/<platform>/dev-requirements.txt
python -m pip install -r requirements/<platform>/sim-requirements.txt
python -m pip install --no-deps -e .
```
Then try to set up one of Elmer or Ansys software working on your device. This is needed so that you can perform one of our simulations on your device, to better understand what format/data the input files for post-processing scripts have, and what file they output. If you get stuck on this part, we can arrange some way to help you get working on the post-processing script tests without a working simulator software.

### Quickly debugging post-processing scripts

A library of simulation scripts in KQC is stored [here](https://github.com/iqm-finland/KQCircuits/tree/main/klayout_package/python/scripts/simulations). Some of these simulation scripts use some of the post-processing scripts, for example [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) uses [produce_epr_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_epr_table.py), [produce_q_factor_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_q_factor_table.py), [elmer_profiler.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/elmer_profiler.py) and [produce_cmatrix_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_cmatrix_table.py). This also happens to be a quick simulation to run, so is a good place to start working on this task. To run it:

```
kqc sim waveguides_sim_cross_section.py
```

> [!NOTE]
> You can modify the [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) script to make it faster by reducing the sweeps. Set `do_solution_sweep = False`, then you can make sure to call the script with option `--n-guides 1`

The `kqc sim` command performs two steps:

- Executes `python klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py`
- Executes `simulation.sh` or `simulation.bat` in the `KQCircuits/tmp/waveguides_sim_cross_section_output` folder

If you open the `simulation.sh` or `simulation.bat` script for editing, you'll see that there are some extra `python` commands for each post-process script after the line: ```echo Post-process```. You can run these individually from your own terminal on the `KQCircuits/tmp/waveguides_sim_cross_section_output` folder to investigate how each post-process script works. You don't need to rerun `kqc sim` command if you want to rerun post-processing scripts. Notice also that at this stage it won't fetch post-processing code from KQC source, but from `KQCircuits/tmp/waveguides_sim_cross_section_output/scripts` where the scripts get copy-pasted from KQC source.

### Available post-processing scripts

Your task is to implement unit tests for 4 post-processing scripts of your choice. We can award you partial bounty if you implement unit tests for less than 4 scripts. Below are descriptions for suitable post-process scripts.

#### [produce_epr_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_epr_table.py)

Used in [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) simulation script.

The script reads the `*_project_results.json` file and some data from e.g. `waveguides_n_guides_1.json` (exported before running Elmer). The arguments were passed to this post-process script, which are read from `produce_epr_table.json` file. The post-process script collects different kinds of data from different sweeps, does some arithmetic calculations, then compiles it into a `*_epr.csv` file.

#### [produce_q_factor_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_q_factor_table.py)

Used in [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) simulation script. Only works correctly if [produce_epr_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_epr_table.py) was run beforehand.

The script reads the `*_project_results.json` file for each sweep, and reads the `*_epr.csv` which was generated by [produce_epr_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_epr_table.py). If latter file does not exist, it will run the script to retrieve it. Performs arithmetic calculations and creates a `*_q_factors.csv` file. The arguments were passed to this post-process script, which are read from `produce_q_factor_table.json` file.

#### [elmer_profiler.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/elmer_profiler.py)

Used in [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) simulation script.

Reads various files from simulation folder and compiles a `*_profile.csv` table containing benchmark data such as runtimes and meshing triangle counts.

#### [produce_cmatrix_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_cmatrix_table.py)

Used in [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) simulation script.

The script reads the `*_project_results.json` file for each sweep, reads `Cs` fields, does some arithmetic operations, then compiles data into `*_results.csv` file. It might read some data from e.g. `waveguides_n_guides_1.json` (exported before running Elmer) to look up fields like `tool`, `deembed_len`, `deembed_cross_section` to perform alternative calculations. Would be good to test edge cases wrt to this behaviour.

#### [produce_z0_table.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/produce_z0_table.py)

You can temporarily add this post-process script to [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) for development purposes:

```
post_process = [
    ...,
    PostProcess("produce_z0_table.py"),
]
```

The script reads the `*_project_results.json` file for each sweep, reads `Cs` and `Ls` fields, does some arithmetic operations, then compiles data into `*_Z0.csv` file.

#### [elmer_cleanup.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/elmer_cleanup.py)

You can temporarily add this post-process script to [waveguides_sim_cross_section.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/waveguides_sim_cross_section.py) for development purposes:

```
post_process = [
    ...,
    PostProcess("elmer_cleanup.py"),
]
```

Deletes certain files from the simulation folder that may get big but don't hold human-readable data, like meshing data. Could be tested easily by:
- set up a temporary folder
- add files intended to be deleted
- add files intended to be persisted
- run post-processing script
- check existence of needed files

#### [interpolate_s_parameters.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/interpolate_s_parameters.py)

Currently only used by [hanger_resonator_sim.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/hanger_resonator_sim.py). You will need a specific version of Elmer that is only available in Linux release, but for Windows can be set up by installing Elmer in WSL using [Singularity container installation guide](https://iqm-finland.github.io/KQCircuits/developer/containers.html#singularity-usage). The `hanger_resonator_sim.py` can take a long time to run, you can try following things to speed up execution just so you can get a working use case of the post process script:

- set bigger `mesh_size = refine_metal_edges(10.0, 1.0)`
- set smaller frequency sweep: `interp_freq = np.linspace(5, 15, 10)`
- multiprocessing features may be buggy especially on Windows environments, set `'gmsh_n_threads': 1`, `'elmer_n_processes': 1`, `'n_workers': 1`

Works best with [polyrat](https://pypi.org/project/polyrat/) library installed in python environment. Reads `sNp` files produced by Elmer, reads requested frequencies from `interpolate_s_parameters.json` file, then interpolates S-parameters at such frequencies from the original `sNp` files, producing `*_interpolated.sNp` files.

#### [calculate_q_from_s.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/calculate_q_from_s.py)

Currently only used by [swissmon_fluxline_sim.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/swissmon_fluxline_sim.py). The post-process script was designed for simulations executed using proprietary Ansys software. Reads a sNp file of form `<SIMNAME>_project_SMatrix.s`, loads the data into [scikit-rf](https://scikit-rf.readthedocs.io/en/latest/) framework, then dumps Q-factor data into ``<SIMNAME>_q.json`` file.

#### [run_pyepr_t1_estimate.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/run_pyepr_t1_estimate.py)

Currently only used by [double_pads_sim.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/double_pads_sim.py). Uses `pyEPR` library that we don't often use anymore, and it is dependent on proprietary Ansys software. Wouldn't recommend working on this script.

#### [rerun_failed_simulations.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/rerun_failed_simulations.py) - NOT NEEDED

This script is quite interactive as it prompts the user about which failed simulations to delete. This might be tricky to unit test by mocking user responses, so we can consider this out-of-scope with this issue.

#### [tls_metal_edge_points.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/tls_metal_edge_points.py) - NOT NEEDED

This script is quite complex, testing this would merit a properly documented issue description in itself - out of scope with this issue.

#### [tls_monte_carlo_points.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/tls_monte_carlo_points.py) - NOT NEEDED

This script is quite complex, testing this would merit a properly documented issue description in itself - out of scope with this issue.

#### [post_process_helpers.py](https://github.com/iqm-finland/KQCircuits/blob/main/klayout_package/python/scripts/simulations/post_process/post_process_helpers.py) - NOT NEEDED

Not a script, a python file with common utility functions. Could be unit tested using standard KQC unit test pipeline, but out of scope with this issue.

### Acceptance criteria

- `pytest` command runs unit tests for scripts under [klayout_package/python/scripts/simulations/post_process](https://github.com/iqm-finland/KQCircuits/tree/main/klayout_package/python/scripts/simulations/post_process)
- Implement a unit test infrastructure for such scripts such that it is easy to maintain and extend such tests, considering that we might create new post-process scripts.
- Implement concrete unit tests for at least 4 post-process scripts.

### PR DESCRIPTION
**Pull request by manasa-manoj-nbr**
Closes #121

## Summary

Implements a pytest infrastructure for testing post-process scripts under `klayout_package/python/scripts/simulations/post_process/`, and adds concrete unit tests for 6 scripts. The focus is on making it easy to maintain and extend tests for future scripts without duplicating boilerplate.

## Changes

**New: `tests/simulations/post_process/conftest.py`**
Shared pytest fixtures used by all post-process tests:
- `sim_folder` — temporary directory mimicking a simulation output folder (chdir + scripts on path)
- `write_simulation(name, definition, results)` — writes paired `<name>.json` / `<name>_project_results.json` mock files
- `run_post_process(script, args)` — runs a script via `runpy` with optional CLI args, the same way `simulation.sh` would
- `read_csv` — parses output CSVs into row dicts

**New: test files for 6 post-process scripts**
| File | Script | Coverage |
|---|---|---|
| `test_elmer_cleanup.py` | `elmer_cleanup.py` | cleans sim folders, skips non-sim dirs, ignores files |
| `test_produce_cmatrix_table.py` | `produce_cmatrix_table.py` | matrix flatten, `Cs` fallback, deembedding edge case |
| `test_produce_z0_table.py` | `produce_z0_table.py` | Z0/c_eff arithmetic, missing-field skip |
| `test_elmer_profiler.py` | `elmer_profiler.py` | gmsh/Elmer runtimes, mesh size, CPU process scaling |
| `test_produce_epr_table.py` | `produce_epr_table.py` | energy participation ratios |
| `test_produce_q_factor_table.py` | `produce_q_factor_table.py` | Q-factor math, chained on real EPR output |

**Fixed: `elmer_cleanup.py` line 30**
`p not in non_sim_dirs` → `p.name not in non_sim_dirs`
The original compared `Path` objects against strings, so the non-sim-dir guard never triggered.

## Checklist

- [x] `pytest tests/simulations/post_process/` passes (11 tests)
- [x] `black --check -l 120` passes on all new files
- [x] `pylint` scores 10.00/10 on all new files
- [x] Copyright headers present on all new files (year 2026)
- [x] No existing tests broken
- [x] Mock data shapes validated against real exported definition files from `kqc sim waveguides_sim_cross_section.py` and the result-serialization code in `elmer_helpers.py` / `cross_section_helpers.py`

## Test results

```
tests/simulations/post_process/test_elmer_cleanup.py ... [ 27%]
tests/simulations/post_process/test_produce_cmatrix_table.py ... [ 54%]
tests/simulations/post_process/test_produce_epr_table.py . [ 63%]
tests/simulations/post_process/test_elmer_profiler.py . [ 72%]
tests/simulations/post_process/test_produce_q_factor_table.py . [ 81%]
tests/simulations/post_process/test_produce_z0_table.py .. [100%]
```
11 passed in 1.65s

## AI disclosure

This PR was developed with Claude assistance for reading the post-process scripts, designing the fixture architecture, and computing expected values in tests. Mock data shapes were validated against real exported simulation files and the result-serialization source code. All changes were manually reviewed and verified for correctness before submission.

### CODE REVIEW

File: tests/simulations/post_process/test_produce_q_factor_table.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# https://www.gnu.org/licenses/gpl-3.0.html.
+#
+# The software distribution should follow IQM trademark policy for open-source software
+# (meetiqm.com/iqm-open-source-trademark-policy). IQM welcomes contributions to the code.
+# Please see our contribution agreements for individuals (meetiqm.com/iqm-individual-contributor-license-agreement)
+# and organizations (meetiqm.com/iqm-organization-contributor-license-agreement).
+
+import json
+
+import pandas as pd
```
[the remark this answers is not in the document]
Comment (manasa-manoj-nbr) [reply tests/simulations/post_process/test_produce_q_factor_table.py]: Switched to the shared read_csv fixture so the test no longer imports pandas directly and stays consistent with the rest of the suite. Fixed.

File: klayout_package/python/scripts/simulations/post_process/elmer_cleanup.py [hunk 2]
```diff
@@ -27,5 +27,5 @@
 
 non_sim_dirs = ["scripts", "log_files", "elmer_data", "s_matrix_plots"]
 for p in path.iterdir():
-    if p.is_dir() and p not in non_sim_dirs:
+    if p.is_dir() and p.name not in non_sim_dirs:
```
[the remark this answers is not in the document]
Comment (manasa-manoj-nbr) [reply klayout_package/python/scripts/simulations/post_process/elmer_cleanup.py:30]: I considered this, but `non_sim_dirs`  is a fixed 4-element list checked against a small number of directories, so the lookup cost is negligible here. I'd prefer to keep this PR's change to `elmer_cleanup.py` limited to the one-line bug fix and leave the script otherwise untouched. Happy to switch it to a set if you'd rather.

**Comment by manasa-manoj-nbr** [comment]
Hi @qpavsmi @iqmtestd , could you please review this when you have a chance? Thank you!

File: klayout_package/python/scripts/simulations/post_process/elmer_cleanup.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (iqmzino) [reply klayout_package/python/scripts/simulations/post_process/elmer_cleanup.py:30]: you can leave as it is. I am fine with the change.

**Comment by qpavsmi** [comment]
I realise now that post process scripts use dependencies that are defined in `sim-requirements.txt` but not in `dev-requirements.txt`, which means that when tests are run in CI they will likely fail due to not finding the dependencies.

If you could make modifications for following files in your PR:

- [init_kqc.sh](https://github.com/iqm-finland/KQCircuits/blob/main/ci/init_kqc.sh) add `-r klayout_package/python/requirements/linux/sim-requirements.txt` in the `python -m pip install` block
- [ci.yaml](https://github.com/iqm-finland/KQCircuits/blob/main/.github/workflows/ci.yaml) in `Tests` and `Tests (non-🐧)` blocks, for `python -m pip install -r` line include `-r klayout_package/python/requirements/???/sim-requirements.txt`, replacing `???` either with `linux` for former block and `${{ matrix.os-folder }}` for latter block.

**Comment by iqmzino** [comment]
@manasa-manoj-nbr  I am happy with the PR and I would approve it once the CI and init_kqc changes are added 
thank you for your work

**Comment by manasa-manoj-nbr** [comment]
Thanks for the review!  I'll address the CI and `init_kqc` changes and update the PR shortly.

**Review by qpavsmi — APPROVED** [review]
Thank you for writing initial tests for the post-process scripts! With this initial testing framework it will be easy for us to add unit tests for subsequent post process scripts that we will add over time

**Comment by manasa-manoj-nbr** [comment]
> Thank you for writing initial tests for the post-process scripts! With this initial testing framework it will be easy for us to add unit tests for subsequent post process scripts that we will add over time

Thanks! I really enjoyed working on this. The test cases are passing now. Can I add a comment on the original issue as well?

**Comment by qpavsmi** [comment]
> > Thank you for writing initial tests for the post-process scripts! With this initial testing framework it will be easy for us to add unit tests for subsequent post process scripts that we will add over time
> 
> Thanks! I really enjoyed working on this. The test cases are passing now. Can I add a comment on the original issue as well?

Yes please!
